dotnet/samples/KernelSyntaxExamples/Example59_OpenAIFunctionCalling.cs

```diff
@@ -0,0 +1,111 @@
+    }
+
+    private static async Task<IKernel> InitializeKernelAsync()
+    {
+        // Create kernel with chat completions service
+        IKernel kernel = new KernelBuilder()
+            .WithLoggerFactory(ConsoleLogger.LoggerFactory)
+            .WithOpenAIChatCompletionService(TestConfiguration.OpenAI.ChatModelId, TestConfiguration.OpenAI.ApiKey, serviceId: "chat")
+            //.WithAzureChatCompletionService(TestConfiguration.AzureOpenAI.ChatDeploymentName, TestConfiguration.AzureOpenAI.Endpoint, TestConfiguration.AzureOpenAI.ApiKey, serviceId: "chat")
+            .Build();
+
+        // Load functions to kernel
+        kernel.ImportFunctions(new TimePlugin(), "TimePlugin");
+        await kernel.ImportPluginFunctionsAsync("KlarnaShoppingPlugin", new Uri("https://www.klarna.com/.well-known/ai-plugin.json"), new OpenApiFunctionExecutionParameters());
+
+        return kernel;
+    }
+
+    private static async Task CompleteChatWithFunctionsAsync(string ask, ChatHistory chatHistory, IChatCompletion chatCompletion, IKernel kernel, OpenAIRequestSettings requestSettings)
+    {
+        Console.WriteLine($"User message: {ask}");
+        chatHistory.AddUserMessage(ask);
+
+        // Send request
+        var chatResult = (await chatCompletion.GetChatCompletionsAsync(chatHistory, requestSettings))[0];
+
+        // Check for message response
+        var chatMessage = await chatResult.GetChatMessageAsync();
+        if (!string.IsNullOrEmpty(chatMessage.Content))
+        {
+            Console.WriteLine(chatMessage.Content);
+
+            // Add the response to chat history
+            chatHistory.AddAssistantMessage(chatMessage.Content);
+        }
+
+        // Check for function response
+        OpenAIFunctionResponse? functionResponse = chatResult.GetFunctionResponse();
+        if (functionResponse is not null)
+        {
+            // Print function response details
+            Console.WriteLine("Function name: " + functionResponse.FunctionName);
+            Console.WriteLine("Plugin name: " + functionResponse.PluginName);
+            Console.WriteLine("Arguments: ");
+            foreach (var parameter in functionResponse.Parameters)
+            {
+                Console.WriteLine($"- {parameter.Key}: {parameter.Value}");
+            }
+
+            // If the function returned by OpenAI is an SKFunction registered with the kernel,
+            // you can invoke it using the following code.
+            if (kernel.Functions.TryGetFunctionAndContext(functionResponse, out ISKFunction? func, out ContextVariables? context))
+            {
+                var result = (await kernel.RunAsync(func, context)).GetValue<string>();
+                if (!string.IsNullOrEmpty(result))
+                {
+                    Console.WriteLine(result);
+
+                    // Add the function result to chat history
+                    chatHistory.AddAssistantMessage(result);
```

Comment:
This should be a 'function' message and then a *new* chat message for the user should be generated.

![image](https://github.com/microsoft/semantic-kernel/assets/35783676/08750a44-ca9b-4ba1-83cf-7d5a2fef24dc)

Reply:
Supporting this requires modifications to our chat message abstractions, here's what it would look like: https://github.com/gitri-ms/semantic-kernel/pull/2. Tracked by #3017 / #1450